### DISCUSSION TITLE
predbat.soc_kw_base10 at 23:30

### DISCUSSION THREAD

**Discussion by mdeakin99** [springfall2008/batpred#1519] [Ideas]
Could we have a sensor that tells me either the predicted SOC% or SOC KWh of the battery just before the next cheap rate kicks in (23:30 for me).  The reason I want to do this is I want to know if I am going to barely make it to 23:30 on my current/predicted usage or if I can put on the washing machine etc which may not be part of the prediction. 

I suggest using BASE10 rather than BEST10  is because BEST10 will try to get to minimum SOC by 23:30 to make the most exporting what's left in the battery.  I think using BASE10 would tell me my SOC at at 23:30 without any dump/pump just before the cheap rate comes in.

In summary,  I am trying to figure out if I can put the washing machine on,  based on the amount of SOC we have left,  so that I have 10% or more at 23:30 (and so the battery will still be supplying the home).

![image](https://github.com/user-attachments/assets/9e22b23e-bc52-44dc-8d30-0d5ef688e349)

**Comment by gcoan** [comment]
You should be able to create a template sensor that gets the kWh SoC from the existing Predbat base10 sensor attributes for any time you want.

![image](https://github.com/user-attachments/assets/fcdf845e-dcc2-4a7b-9ef2-dc1cbb4a9f54)

You'll need to iterate across the 'results' attribute to find the right time you want (Google search for the template YAML ..)

**Comment by mdeakin99** [comment]
Indeed.  Google and ChatGPT tried but couldn’t get the result I was looking for.

Thought it might be something others without inverter control might want to see:

Mark

On 10 Oct 2024, at 10:57, Geoffrey Coan ***@***.***> wrote:

﻿

You should be able to create a template sensor that gets the kWh SoC from the existing Predbat base10 sensor attributes for any time you want.

E06859A1-C4FB-4D39-B3E6-33DADA3695D2.jpeg (view on web)<https://github.com/user-attachments/assets/fcdf845e-dcc2-4a7b-9ef2-dc1cbb4a9f54>

You'll need to iterate across the 'results' attribute to find the right time you want (Google search for the template YAML ..)

**Reply by gcoan** [reply]
You didn't try hard enough 😉

{% set x=state_attr('predbat.soc_kw_base10', 'results') %}

Will get the results attribute

{{ x }} will show you the contents of the results, 
e.g. 

{'2024-10-10T19:50:00+0100': 8.18, '2024-10-10T19:55:00+0100': 8.062, '2024-10-10T20:00:00+0100': 7.808, '2024-10-10T20:05:00+0100': 7.786, '2024-10-10T20:10:00+0100': 7.582, '2024-10-10T20:15:00+0100': 7.537, '2024-10-10T20:20:00+0100': 7.333, '2024-10-10T20:25:00+0100': 7.315, '2024-10-10T20:30:00+0100': 7.163, '2024-10-10T20:35:00+0100': 7.151, '2024-10-10T20:40:00+0100': 7.007, '2024-10-10T20:45:00+0100': 6.978, '2024-10-10T20:50:00+0100': 6.803, '2024-10-10T20:55:00+0100': 6.787, '2024-10-10T21:00:00+0100': 6.632, '2024-10-10T21:05:00+0100': 6.615, '2024-10-10T21:10:00+0100': 6.501, '2024-10-10T21:15:00+0100': 6.492, '2024-10-10T21:20:00+0100': 6.403, '2024-10-10T21:25:00+0100': 6.366, ... }

As you can see its a set of key/value pairs, and

{{ x['2024-10-10T23:30:00+0100'] }} 

Will give you the SoC forecast in kWh at 11:30pm tonight.

You'll need to do a bit of extra coding to generate tonight's date & time in the correct format, allowing for UTC/BST, but this should get you going.

I do recommend watching the Jinja course from Smart Home Junkie on YouTube, he builds up bit by bit how to do loops, read sensors, etc

**Comment by Nilogax** [comment]
Here's an example for min charge just before the cheap rate Flux window - you should be able to adapt this 

```jinja
{% set capacity = states('sensor.givtcp_SERIAL_battery_capacity_kwh')|  float  %}
{% set tomorrow0200 = (today_at('02:00') + timedelta(days=1)).strftime ('%Y-%m-%dT%H:%M:%S%z' )
    %}
{% set tomorrow0155 = (today_at('01:55') + timedelta(days=1)).strftime ('%Y-%m-%dT%H:%M:%S%z' )
    %}
{% if state_attr('predbat.soc_kw_best','results') [tomorrow0200] is defined %}
{{(100 * state_attr('predbat.soc_kw_best','results') [tomorrow0200] / capacity) }}
{% else %}
{{(100 * state_attr('predbat.soc_kw_best','results') [tomorrow0155] / capacity) }}
{% endif %}
```
